routing/result_interpretation.go

```diff
@@ -517,11 +517,20 @@ func (i *interpretedResult) processPaymentOutcomeIntermediate(route *mcRoute,
 		if introIdx == len(route.hops.Val)-1 {
 			i.finalFailureReason = &reasonError
 		} else {
-			// If there are other hops between the recipient and
-			// introduction node, then we just penalize the last
-			// hop in the blinded route to minimize the storage of
-			// results for ephemeral keys.
-			i.failPairBalance(route, len(route.hops.Val)-1)
+			// Penalize all hops after the introduction node up to
+			// the final recipient. We cannot just penalize the
+			// last hop because when finding a path to a blinded
+			// route we swap the target pubkey for a general one
+			// so that multi-path payment can be attempted. However
```

Comment:
it isnt super clear to me from the comment what the issue is/was. Is it that we only want to penalise one path instead of all the paths? why not just use the real (pre swap) final key & penalise that?

Reply:
We currently do use the real-path, the problem arises that we will not use this information during pathfinding hence we will try to reattempt paths which already failed with the invalid blinding error. So that's why I propose penalizing the whole path which makes it bullet-proof.

Reply:
I don't see the problem yet, could you specify a reproduction scenario (and logs with amounts)? The last hop (swapped) pubkey should be stable over a payment, right? So penalizing the last hop pair should be in effect even if further partial attempts are made. I can see though that the amount will be split often and the route will be re-attempted until the min shard size is reached

Reply:
the setup is pretty simple, just create a blinded path where the payment fails in the blinded path. LND will retry the payment until timeout because the wrong pair is penalized in the MC.

Reply:
LND will also not split the amount because it always find the same route.

Reply:
ok, yes I see it now thanks for the examples, in pathfinding we use the swapped pubkey T for Z (X, T), but when reporting to mission control we use the original route and therefore we punish (X, Z). I think we can fix it like it is done in this PR.

However, as noted in the comments in the lines above this change, we would fail a payment that has for example two single-hop hop hints after we fail to send over one of them. As noted in the comment, this was ok for non-mpp scenarios, but we support mpp now. So I think the cleanest fix would be to add a dummy hop to the single-hop hints such that we have an extension like [(X,Z), (Z,T)] and a search a path to T. If we make T static, we could even reuse those mc reports for the same blinded route (not sure if there's an application). The dummy hop would need to be removed from the route after reporting to mission control.

Reply:
I tested the current setup for a multipart payment, but we need something more, because with `failPairBalance` we actually take the previous hop's `amtToFwd`, which is set to zero for blinded hops. This leads to an over-penalization, we basically say that we can't forward anything by setting a failure amount of zero. To fix that, I'd suggest to add and use:
```go
// failBlindedRoute marks a blinded route as failed for the specific amount to
// send.
func (i *interpretedResult) failBlindedRoute(rt *mcRoute) {
	// We fail the last pair of the route, in order to fail the complete
	// blinded route. This is because the combination of ephemeral pubkeys
	// is unique to the route. We fail the last pair in order to not punish
	// the introduction node, since we don't want to disincentivize them
	// from providing that service.
	pair, _ := getPair(rt, len(rt.hops.Val)-1)

	// Since all the hops along a blinded path don't have any amount set, we
	// extract the minimal amount to punish from the value that is tried to
	// be sent to the receiver.
	amt := rt.hops.Val[len(rt.hops.Val)-1].amtToFwd.Val

	i.pairResults[pair] = failPairResult(amt)
}
```